include/cuda/memory_resource

```diff
@@ -0,0 +1,809 @@
+   * host), or behavior is undefined.
+   *
+   * \param __p Pointer to storage to be deallocated.
+   * \param __bytes The size in bytes of the allocation. This must be equal to
+   * the value of `__bytes` that was specified to the `allocate` or
+   * `allocate_async` call that returned `__p`.
+   * \param __stream The stream on which to perform the deallocation.
+   */
+  void deallocate_async(void *__mem, size_t __bytes, stream_view __stream) {
+    deallocate_async(__mem, __bytes, default_alignment, __stream);
+  }
+
+  /*!
+   * \brief Deallocates the storage pointed to by `__p` in stream order on
+   * `__stream`.
+   *
+   * `__p` must have been returned by a prior call to
+   * `allocate_async(__bytes, __alignment)` or `allocate(__bytes,
+   * __alignment)` on a `stream_ordered_memory_resource` that compares
+   * equal to `*this`, and the storage it points to must not yet have been
+   * deallocated, otherwise behavior is undefined.
+   *
+   * Asynchronous, stream-ordered operations on `__stream` initiated before
+   * `deallocate_async(__p, __bytes, __stream)` may still access the storage
+   * pointed to by `__p` after `deallocate_async` returns.
+   *
+   * Storage deallocated on `__stream` may be reused by a future
+   * call to `allocate_async` on the same stream without synchronizing
+   * `__stream`. Therefore,  `__stream` is typically the last stream on which
+   * `__p` was last used. It is the caller's responsibility to ensure the
+   * storage pointed to by `__p` is not in use on any other stream (or the
+   * host), or behavior is undefined.
+   *
+   * \param __p Pointer to storage to be deallocated.
+   * \param __bytes The size in bytes of the allocation. This must be equal to
+   * the value of `__bytes` that was specified to the `allocate` or
+   * `allocate_async` call that returned `__p`.
+   * \param __alignment The alignment of the allocation. This must be equal to
+   * the value of `__alignment` that was specified to the `allocate` or
+   * `allocate_async` call that returned `__p`.
+   * \param __stream The stream on which to perform the deallocation.
+   */
+  void deallocate_async(void *__mem, size_t __bytes, size_t __alignment, stream_view __stream) {
+    do_deallocate_async(__mem, __bytes, __alignment, __stream);
+  }
+  virtual void *do_allocate_async(size_t __bytes, size_t __alignment, stream_view __stream) = 0;
+  virtual void do_deallocate_async(void *__mem, size_t __bytes, size_t __alignment, stream_view __stream) = 0;
+};
+
+/*!
+ * \brief Abstract interface for context specific memory allocation.
+ *
+ * \tparam _MemoryKind The kind of the allocated memory.
+ * \tparam _Context The execution context on which the storage may be used
+ * without synchronization
+ */
+template <typename _MemoryKind, typename _Context = any_context>
+class memory_resource : private virtual memory_resource_base, private detail::__get_context_impl<_Context> {
+public:
+  using memory_kind = _MemoryKind;
```

Comment:
Wasn't there an issue with name collision of using `memory_kind` for the typedef here? I recall I just used `kind`.